Index: vdr-1.7.28/menu.h
===================================================================
--- vdr-1.7.28.orig/menu.h	2012-06-07 12:20:45.835592267 +0200
+++ vdr-1.7.28/menu.h	2012-06-07 12:20:45.951592272 +0200
@@ -299,6 +299,7 @@
   virtual void Hide(void);
   bool Visible(void) { return visible; }
   static void SetRecording(const char *FileName);
+  static void SetRecording(const char *FileName, const char *) { SetRecording(FileName); }
   static const char *NowReplaying(void);
   static const char *LastReplayed(void);
   static void ClearLastReplayed(const char *FileName);
